File: src/apps/business-os-mobile/ios/Sources/CTOXMobileCore/OfficePack.swift
import CryptoKit
import Foundation

public struct OfficePackFile: Codable, Equatable, Sendable { public let path: String; public let size: Int64; public let sha256: String }
public struct OfficePackManifest: Codable, Equatable, Sendable {
    public let format: String
    public let packID: String
    public let sourceRevision: String
    public let appVersion: String
    public let totalBytes: Int64
    public let files: [OfficePackFile]
    enum CodingKeys: String, CodingKey { case format, files; case packID = "pack_id", sourceRevision = "source_revision", appVersion = "app_version", totalBytes = "total_bytes" }
}

public enum OfficePackError: Error, Equatable, Sendable { case manifest, revision, appVersion, canceled, missing(String), size(String), hash(String), total }
public enum OfficePackVerifier {
    public static func verify(root: URL, manifest: OfficePackManifest, sourceRevision: String, appVersion: String, canceled: () -> Bool = { false }, progress: (Double) -> Void = { _ in }) throws {
        guard manifest.format == "ctox.mobile.shell-pack.v1", manifest.packID == "ctox-office" else { throw OfficePackError.manifest }
        guard manifest.sourceRevision == sourceRevision else { throw OfficePackError.revision }
        guard manifest.appVersion == appVersion else { throw OfficePackError.appVersion }
        let canonicalRoot = root.standardizedFileURL.path + "/"
        var verified: Int64 = 0
        for file in manifest.files {
            if canceled() { throw OfficePackError.canceled }
            guard !file.path.hasPrefix("/"), !file.path.split(separator: "/").contains("..") else { throw OfficePackError.manifest }
            let url = root.appending(path: file.path).standardizedFileURL
            guard url.path.hasPrefix(canonicalRoot), let data = try? Data(contentsOf: url) else { throw OfficePackError.missing(file.path) }
            guard data.count == file.size else { throw OfficePackError.size(file.path) }
            let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
            guard digest == file.sha256 else { throw OfficePackError.hash(file.path) }
            verified += file.size
            progress(manifest.totalBytes > 0 ? Double(verified) / Double(manifest.totalBytes) : 1)
        }
        guard verified == manifest.totalBytes else { throw OfficePackError.total }
    }
}

File: src/apps/business-os-mobile/shared/office-pack.mjs
import crypto from "node:crypto";
import fs from "node:fs/promises";
import path from "node:path";
import { PACK_FORMAT, PACK_ID } from "./constants.mjs";

export async function sha256File(file) {
  const bytes = await fs.readFile(file);
  return crypto.createHash("sha256").update(bytes).digest("hex");
}

export async function buildPackManifest(root, { sourceRevision, appVersion }) {
  const files = [];
  async function walk(relative = "") {
    const dir = path.join(root, relative);
    for (const entry of (await fs.readdir(dir, { withFileTypes: true })).sort((a, b) => a.name.localeCompare(b.name))) {
      const child = path.posix.join(relative.split(path.sep).join(path.posix.sep), entry.name);
      const absolute = path.join(root, child);
      if (entry.isDirectory()) await walk(child);
      else if (entry.isFile()) {
        const stat = await fs.stat(absolute);
        files.push({ path: child, size: stat.size, sha256: await sha256File(absolute) });
      }
    }
  }
  await walk();
  return {
    format: PACK_FORMAT,
    pack_id: PACK_ID,
    source_revision: sourceRevision,
    app_version: appVersion,
    total_bytes: files.reduce((sum, file) => sum + file.size, 0),
    files,
  };
}

export async function verifyPack(root, manifest, { sourceRevision, appVersion, signal, onProgress = () => {} } = {}) {
  if (manifest?.format !== PACK_FORMAT || manifest?.pack_id !== PACK_ID) throw new Error("unsupported office pack manifest");
  if (manifest.source_revision !== sourceRevision) throw new Error("office pack revision mismatch");
  if (manifest.app_version !== appVersion) throw new Error("office pack app version mismatch");
  let verified = 0;
  for (const file of manifest.files || []) {
    if (signal?.aborted) throw new Error("office pack activation canceled");
    if (!file.path || path.isAbsolute(file.path) || file.path.includes("..")) throw new Error("office pack path is invalid");
    const absolute = path.resolve(root, file.path);
    const canonicalRoot = path.resolve(root);
    if (!absolute.startsWith(`${canonicalRoot}${path.sep}`)) throw new Error("office pack path escapes root");
    const stat = await fs.stat(absolute).catch(() => null);
    if (!stat?.isFile() || stat.size !== file.size) throw new Error(`office pack file size mismatch: ${file.path}`);
    if (await sha256File(absolute) !== file.sha256) throw new Error(`office pack hash mismatch: ${file.path}`);
    verified += file.size;
    onProgress(manifest.total_bytes ? verified / manifest.total_bytes : 1);
  }
  if (verified !== manifest.total_bytes) throw new Error("office pack total byte mismatch");
  return true;
}
